File: tests/test_geowave_kat.py
#!/usr/bin/env python3
"""
Geometric Waveform Cipher Known-Answer Tests (KATs)
Test vectors for validating geometric waveform hashing and encryption
"""

import sys
import os
import hashlib
import json
from typing import List, Dict, Any, Tuple

# Add core modules to path
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', 'core'))

try:
    from encryption.geometric_waveform_hash import (
        geometric_waveform_hash,
        generate_waveform_hash,
        verify_waveform_hash
    )
    from encryption.resonance_encrypt import resonance_encrypt
    GEOWAVE_AVAILABLE = True
except ImportError:
    GEOWAVE_AVAILABLE = False

class GeometricWaveformKAT:
    """Known-Answer Tests for Geometric Waveform Cipher"""
    
    def __init__(self):
        self.test_results = []
        self.known_test_vectors = self.generate_known_test_vectors()
        
    def generate_known_test_vectors(self) -> List[Dict[str, Any]]:
        """Generate known test vectors for validation"""
        test_vectors = []
        
        # Test Vector 1: Simple sine wave
        test_vectors.append({
            'name': 'sine_wave_8',
            'input': [0.0, 0.707, 1.0, 0.707, 0.0, -0.707, -1.0, -0.707],
            'expected_amplitude': 0.707,
            'expected_phase_range': (0.0, 1.0),
            'description': 'Simple sine wave with 8 samples'
        })
        
        # Test Vector 2: Delta function
        test_vectors.append({
            'name': 'delta_function',
            'input': [1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0],
            'expected_amplitude': 0.125,
            'expected_phase_range': (0.0, 1.0),
            'description': 'Delta function (impulse)'
        })
        
        # Test Vector 3: Step function
        test_vectors.append({
            'name': 'step_function',
            'input': [0.0, 0.0, 0.0, 0.0, 1.0, 1.0, 1.0, 1.0],
            'expected_amplitude': 0.5,
            'expected_phase_range': (0.0, 1.0),
            'description': 'Step function'
        })
        
        # Test Vector 4: Constant signal
        test_vectors.append({
            'name': 'constant_signal',
            'input': [0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5],
            'expected_amplitude': 0.5,
            'expected_phase_range': (0.0, 1.0),
            'description': 'Constant signal'
        })
        
        # Test Vector 5: Linear ramp
        test_vectors.append({
            'name': 'linear_ramp',
            'input': [0.0, 0.125, 0.25, 0.375, 0.5, 0.625, 0.75, 0.875],
            'expected_amplitude': 0.4375,
            'expected_phase_range': (0.0, 1.0),
            'description': 'Linear ramp'
        })
        
        # Test Vector 6: Alternating pattern
        test_vectors.append({
            'name': 'alternating_pattern',
            'input': [1.0, -1.0, 1.0, -1.0, 1.0, -1.0, 1.0, -1.0],
            'expected_amplitude': 1.0,
            'expected_phase_range': (0.0, 1.0),
            'description': 'Alternating +1/-1 pattern'
        })
        
        # Test Vector 7: Complex waveform
        test_vectors.append({
            'name': 'complex_waveform',
            'input': [0.5, 0.8, 0.3, -0.2, 0.9, -0.4, 0.1, 0.6],
            'expected_amplitude': 0.45,
            'expected_phase_range': (0.0, 1.0),
            'description': 'Complex waveform with mixed amplitudes'
        })
        
        # Test Vector 8: Symmetric waveform
        test_vectors.append({
            'name': 'symmetric_waveform',
            'input': [0.1, 0.3, 0.7, 0.9, 0.9, 0.7, 0.3, 0.1],
            'expected_amplitude': 0.5,
            'expected_phase_range': (0.0, 1.0),
            'description': 'Symmetric waveform'
        })
        
        return test_vectors
    
    def test_geometric_hash_consistency(self):
        """Test that geometric hashing produces consistent results"""
        if not GEOWAVE_AVAILABLE:
            self.test_results.append({
                'test': 'geometric_hash_consistency',
                'status': 'SKIPPED',
                'reason': 'Geometric waveform modules not available'
            })
            return
        
        for test_vector in self.known_test_vectors:
            try:
                waveform = test_vector['input']
                name = test_vector['name']
                
                # Generate hash multiple times
                hash1 = geometric_waveform_hash(waveform)
                hash2 = geometric_waveform_hash(waveform)
                hash3 = geometric_waveform_hash(waveform)
                
                # All hashes should be identical
                consistent = hash1 == hash2 == hash3
                
                self.test_results.append({
                    'test': f'geometric_hash_consistency_{name}',
                    'status': 'PASS' if consistent else 'FAIL',
                    'hash1': hash1,
                    'hash2': hash2,
                    'hash3': hash3,
                    'consistent': consistent,
                    'description': test_vector['description']
                })
                
            except Exception as e:
                self.test_results.append({
                    'test': f'geometric_hash_consistency_{name}',
                    'status': 'ERROR',
                    'error': str(e),
                    'description': test_vector['description']
                })
    
    def test_hash_uniqueness(self):
        """Test that different waveforms produce different hashes"""
        if not GEOWAVE_AVAILABLE:
            self.test_results.append({
                'test': 'hash_uniqueness',
                'status': 'SKIPPED',
                'reason': 'Geometric waveform modules not available'
            })
            return
        
        try:
            # Generate hashes for all test vectors
            hashes = []
            for test_vector in self.known_test_vectors:
                waveform = test_vector['input']
                hash_value = geometric_waveform_hash(waveform)
                hashes.append(hash_value)
            
            # Check for uniqueness
            unique_hashes = set(hashes)
            all_unique = len(unique_hashes) == len(hashes)
            
            self.test_results.append({
                'test': 'hash_uniqueness',
                'status': 'PASS' if all_unique else 'FAIL',
                'total_hashes': len(hashes),
                'unique_hashes': len(unique_hashes),
                'collision_count': len(hashes) - len(unique_hashes),
                'hashes': hashes
            })
            
        except Exception as e:
            self.test_results.append({
                'test': 'hash_uniqueness',
                'status': 'ERROR',
                'error': str(e)
            })
    
    def test_waveform_hash_format(self):
        """Test that hash output format is correct"""
        if not GEOWAVE_AVAILABLE:
            self.test_results.append({
                'test': 'waveform_hash_format',
                'status': 'SKIPPED',
                'reason': 'Geometric waveform modules not available'
            })
            return
        
        for test_vector in self.known_test_vectors:
            try:
                waveform = test_vector['input']
                name = test_vector['name']
                
                hash_value = geometric_waveform_hash(waveform)
                
                # Check hash format
                format_tests = {
                    'is_string': isinstance(hash_value, str),
                    'has_amplitude_prefix': hash_value.startswith('A'),
                    'has_phase_component': '_P' in hash_value,
                    'has_hex_suffix': len(hash_value) > 20,  # Should have hex hash
                    'correct_length': len(hash_value) > 70  # A + 4 digits + _P + 4 digits + _ + 64 hex chars
                }
                
                all_format_checks_pass = all(format_tests.values())
                
                self.test_results.append({
                    'test': f'waveform_hash_format_{name}',
                    'status': 'PASS' if all_format_checks_pass else 'FAIL',
                    'hash_value': hash_value,
                    'hash_length': len(hash_value),
                    'format_checks': format_tests,
                    'description': test_vector['description']
                })
                
            except Exception as e:
                self.test_results.append({
                    'test': f'waveform_hash_format_{name}',
                    'status': 'ERROR',
                    'error': str(e),
                    'description': test_vector['description']
                })
    
    def test_golden_ratio_properties(self):
        """Test that golden ratio optimization is applied correctly"""
        if not GEOWAVE_AVAILABLE:
            self.test_results.append({
                'test': 'golden_ratio_properties',
                'status': 'SKIPPED',
                'reason': 'Geometric waveform modules not available'
            })
            return
        
        try:
            # Golden ratio
            phi = (1 + (5 ** 0.5)) / 2
            
            # Test with simple waveform
            waveform = [0.5, 0.8, 0.3, 0.1, 0.9, 0.2, 0.7, 0.4]
            hash_value = geometric_waveform_hash(waveform)
            
            # Extract amplitude and phase from hash
            parts = hash_value.split('_')
            if len(parts) >= 2:
                amp_str = parts[0][1:]  # Remove 'A' prefix
                phase_str = parts[1][1:]  # Remove 'P' prefix
                
                try:
                    amplitude = float(amp_str)
                    phase = float(phase_str)
                    
                    # Check if values are within [0,1] range (normalized)
                    amp_in_range = 0.0 <= amplitude <= 1.0
                    phase_in_range = 0.0 <= phase <= 1.0
                    
                    # Check if golden ratio influence is present
                    # (This is a heuristic test - actual implementation details may vary)
                    golden_ratio_test = amp_in_range and phase_in_range
                    
                    self.test_results.append({
                        'test': 'golden_ratio_properties',
                        'status': 'PASS' if golden_ratio_test else 'FAIL',
                        'golden_ratio': phi,
                        'extracted_amplitude': amplitude,
                        'extracted_phase': phase,
                        'amplitude_in_range': amp_in_range,
                        'phase_in_range': phase_in_range,
                        'hash_value': hash_value
                    })
                    
                except ValueError:
                    self.test_results.append({
                        'test': 'golden_ratio_properties',
                        'status': 'FAIL',
                        'error': 'Could not extract amplitude/phase from hash',
                        'hash_value': hash_value
                    })
            else:
                self.test_results.append({
                    'test': 'golden_ratio_properties',
                    'status': 'FAIL',
                    'error': 'Hash format does not match expected pattern',
                    'hash_value': hash_value
                })
                
        except Exception as e:
            self.test_results.append({
                'test': 'golden_ratio_properties',
                'status': 'ERROR',
                'error': str(e)
            })
    
    def test_empty_waveform_handling(self):
        """Test handling of edge cases like empty waveforms"""
        if not GEOWAVE_AVAILABLE:
            self.test_results.append({
                'test': 'empty_waveform_handling',
                'status': 'SKIPPED',
                'reason': 'Geometric waveform modules not available'
            })
            return
        
        edge_cases = [
            ('empty_list', []),
            ('single_value', [0.5]),
            ('two_values', [0.3, 0.7]),
            ('all_zeros', [0.0, 0.0, 0.0, 0.0]),
            ('all_ones', [1.0, 1.0, 1.0, 1.0])
        ]
        
        for case_name, waveform in edge_cases:
            try:
                hash_value = geometric_waveform_hash(waveform)
                
                # Should handle gracefully without crashing
                handles_gracefully = hash_value is not None and isinstance(hash_value, str)
                
                self.test_results.append({
                    'test': f'empty_waveform_handling_{case_name}',
                    'status': 'PASS' if handles_gracefully else 'FAIL',
                    'input_waveform': waveform,
                    'hash_value': hash_value,
                    'handles_gracefully': handles_gracefully
                })
                
            except Exception as e:
                self.test_results.append({
                    'test': f'empty_waveform_handling_{case_name}',
                    'status': 'ERROR',
                    'input_waveform': waveform,
                    'error': str(e)
                })
    
    def run_all_tests(self):
        """Run all known-answer tests"""
        print("Running Geometric Waveform Cipher Known-Answer Tests...")
        print("=" * 60)
        
        # Run all test methods
        self.test_geometric_hash_consistency()
        self.test_hash_uniqueness()
        self.test_waveform_hash_format()
        self.test_golden_ratio_properties()
        self.test_empty_waveform_handling()
        
        # Calculate summary statistics
        total_tests = len(self.test_results)
        passed_tests = sum(1 for result in self.test_results if result['status'] == 'PASS')
        failed_tests = sum(1 for result in self.test_results if result['status'] == 'FAIL')
        error_tests = sum(1 for result in self.test_results if result['status'] == 'ERROR')
        skipped_tests = sum(1 for result in self.test_results if result['status'] == 'SKIPPED')
        
        print(f"Test Results Summary:")
        print(f"  Total: {total_tests}")
        print(f"  Passed: {passed_tests}")
        print(f"  Failed: {failed_tests}")
        print(f"  Errors: {error_tests}")
        print(f"  Skipped: {skipped_tests}")
        
        # Print individual test results
        print("\nDetailed Results:")
        for result in self.test_results:
            status = result['status']
            test_name = result['test']
            print(f"  {test_name}: {status}")
            
            if status == 'ERROR':
                print(f"    Error: {result['error']}")
            elif status == 'SKIPPED':
                print(f"    Reason: {result['reason']}")
        
        # Save results to JSON file
        with open('geowave_kat_results.json', 'w') as f:
            json.dump({
                'summary': {
                    'total': total_tests,
                    'passed': passed_tests,
                    'failed': failed_tests,
                    'errors': error_tests,
                    'skipped': skipped_tests,
                    'success_rate': passed_tests / total_tests if total_tests > 0 else 0
                },
                'test_vectors': self.known_test_vectors,
                'results': self.test_results
            }, f, indent=2)
        
        print(f"\nResults saved to: geowave_kat_results.json")
        
        # Return success if all non-skipped tests passed
        return failed_tests == 0 and error_tests == 0

def main():
    """Run geometric waveform cipher KATs"""
    tester = GeometricWaveformKAT()
    success = tester.run_all_tests()
    
    if success:
        print("\n✅ All geometric waveform cipher KATs passed!")
        return 0
    else:
        print("\n❌ Some geometric waveform cipher KATs failed!")
        return 1

if __name__ == "__main__":
    sys.exit(main())

File: quantoniumos/tests/test_geowave_kat.cpp
/*
 * Geometric Waveform Cipher C++ Known-Answer Tests
 * Patent-protected geometric waveform hashing validation
 * USPTO Application #19/169,399
 */

#include <iostream>
#include <vector>
#include <string>
#include <cmath>
#include <cassert>
#include <sstream>
#include <iomanip>
#include <map>
#include <set>

// Golden ratio constant
const double PHI = (1.0 + std::sqrt(5.0)) / 2.0;

class GeometricWaveformCipher {
private:
    std::vector<double> waveform;
    double amplitude;
    double phase;
    
public:
    GeometricWaveformCipher(const std::vector<double>& input) : waveform(input) {
        calculateAmplitudePhase();
    }
    
    void calculateAmplitudePhase() {
        if (waveform.empty()) {
            amplitude = 0.0;
            phase = 0.0;
            return;
        }
        
        // Calculate amplitude as average absolute value
        double sum = 0.0;
        for (double val : waveform) {
            sum += std::abs(val);
        }
        amplitude = sum / waveform.size();
        
        // Calculate phase using geometric principles
        double even_sum = 0.0, odd_sum = 0.0;
        for (size_t i = 0; i < waveform.size(); ++i) {
            if (i % 2 == 0) {
                even_sum += waveform[i];
            } else {
                odd_sum += waveform[i];
            }
        }
        
        phase = std::atan2(odd_sum, even_sum) / (2.0 * M_PI);
        phase = (phase + 1.0) / 2.0; // Normalize to [0,1]
        
        // Apply golden ratio optimization
        amplitude = std::fmod(amplitude * PHI, 1.0);
        phase = std::fmod(phase * PHI, 1.0);
    }
    
    std::string generateHash() const {
        std::stringstream ss;
        ss << "A" << std::fixed << std::setprecision(4) << amplitude
           << "_P" << std::fixed << std::setprecision(4) << phase;
        
        // Generate waveform samples for hash
        std::vector<double> samples;
        for (int i = 0; i < 64; ++i) {
            double t = i / 64.0;
            double value = amplitude * std::sin(2.0 * M_PI * t + phase * 2.0 * M_PI);
            samples.push_back(std::round(value * 1000000.0) / 1000000.0); // 6 decimal precision
        }
        
        // Simple hash function (in real implementation, use SHA-256)
        std::hash<std::string> hasher;
        std::string sample_str;
        for (double sample : samples) {
            sample_str += std::to_string(sample) + "_";
        }
        
        size_t hash_value = hasher(sample_str);
        ss << "_" << std::hex << hash_value;
        
        return ss.str();
    }
    
    bool verifyHash(const std::string& hash) const {
        return hash == generateHash();
    }
    
    double getAmplitude() const { return amplitude; }
    double getPhase() const { return phase; }
};

// Known-Answer Test vectors
struct TestVector {
    std::string name;
    std::vector<double> input;
    std::string description;
};

class GeometricWaveformKAT {
private:
    std::vector<TestVector> test_vectors;
    std::map<std::string, bool> results;
    
public:
    GeometricWaveformKAT() {
        initializeTestVectors();
    }
    
    void initializeTestVectors() {
        test_vectors = {
            {"sine_wave_8", {0.0, 0.707, 1.0, 0.707, 0.0, -0.707, -1.0, -0.707}, "Simple sine wave"},
            {"delta_function", {1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0}, "Delta function"},
            {"step_function", {0.0, 0.0, 0.0, 0.0, 1.0, 1.0, 1.0, 1.0}, "Step function"},
            {"constant_signal", {0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5}, "Constant signal"},
            {"linear_ramp", {0.0, 0.125, 0.25, 0.375, 0.5, 0.625, 0.75, 0.875}, "Linear ramp"},
            {"alternating", {1.0, -1.0, 1.0, -1.0, 1.0, -1.0, 1.0, -1.0}, "Alternating pattern"},
            {"complex_wave", {0.5, 0.8, 0.3, -0.2, 0.9, -0.4, 0.1, 0.6}, "Complex waveform"},
            {"symmetric", {0.1, 0.3, 0.7, 0.9, 0.9, 0.7, 0.3, 0.1}, "Symmetric waveform"}
        };
    }
    
    void testHashConsistency() {
        std::cout << "Testing hash consistency..." << std::endl;
        
        for (const auto& tv : test_vectors) {
            GeometricWaveformCipher cipher(tv.input);
            
            std::string hash1 = cipher.generateHash();
            std::string hash2 = cipher.generateHash();
            std::string hash3 = cipher.generateHash();
            
            bool consistent = (hash1 == hash2) && (hash2 == hash3);
            results["hash_consistency_" + tv.name] = consistent;
            
            std::cout << "  " << tv.name << ": " << (consistent ? "PASS" : "FAIL") << std::endl;
            if (!consistent) {
                std::cout << "    Hash1: " << hash1 << std::endl;
                std::cout << "    Hash2: " << hash2 << std::endl;
                std::cout << "    Hash3: " << hash3 << std::endl;
            }
        }
    }
    
    void testHashUniqueness() {
        std::cout << "Testing hash uniqueness..." << std::endl;
        
        std::vector<std::string> hashes;
        for (const auto& tv : test_vectors) {
            GeometricWaveformCipher cipher(tv.input);
            hashes.push_back(cipher.generateHash());
        }
        
        std::set<std::string> unique_hashes(hashes.begin(), hashes.end());
        bool all_unique = (unique_hashes.size() == hashes.size());
        results["hash_uniqueness"] = all_unique;
        
        std::cout << "  Uniqueness: " << (all_unique ? "PASS" : "FAIL") << std::endl;
        std::cout << "  Total hashes: " << hashes.size() << std::endl;
        std::cout << "  Unique hashes: " << unique_hashes.size() << std::endl;
    }
    
    void testGoldenRatioOptimization() {
        std::cout << "Testing golden ratio optimization..." << std::endl;
        
        for (const auto& tv : test_vectors) {
            GeometricWaveformCipher cipher(tv.input);
            
            double amplitude = cipher.getAmplitude();
            double phase = cipher.getPhase();
            
            bool amp_in_range = (amplitude >= 0.0 && amplitude <= 1.0);
            bool phase_in_range = (phase >= 0.0 && phase <= 1.0);
            
            bool golden_ratio_applied = amp_in_range && phase_in_range;
            results["golden_ratio_" + tv.name] = golden_ratio_applied;
            
            std::cout << "  " << tv.name << ": " << (golden_ratio_applied ? "PASS" : "FAIL") << std::endl;
            std::cout << "    Amplitude: " << amplitude << std::endl;
            std::cout << "    Phase: " << phase << std::endl;
        }
    }
    
    void testHashVerification() {
        std::cout << "Testing hash verification..." << std::endl;
        
        for (const auto& tv : test_vectors) {
            GeometricWaveformCipher cipher(tv.input);
            
            std::string hash = cipher.generateHash();
            bool verified = cipher.verifyHash(hash);
            results["hash_verification_" + tv.name] = verified;
            
            std::cout << "  " << tv.name << ": " << (verified ? "PASS" : "FAIL") << std::endl;
        }
    }
    
    void runAllTests() {
        std::cout << "=== Geometric Waveform Cipher C++ Known-Answer Tests ===" << std::endl;
        std::cout << "Patent-protected algorithms - USPTO Application #19/169,399" << std::endl;
        std::cout << "Golden ratio φ = " << PHI << std::endl;
        std::cout << std::endl;
        
        testHashConsistency();
        std::cout << std::endl;
        
        testHashUniqueness();
        std::cout << std::endl;
        
        testGoldenRatioOptimization();
        std::cout << std::endl;
        
        testHashVerification();
        std::cout << std::endl;
        
        // Summary
        int total_tests = results.size();
        int passed_tests = 0;
        for (const auto& result : results) {
            if (result.second) passed_tests++;
        }
        
        std::cout << "=== Test Summary ===" << std::endl;
        std::cout << "Total tests: " << total_tests << std::endl;
        std::cout << "Passed tests: " << passed_tests << std::endl;
        std::cout << "Failed tests: " << (total_tests - passed_tests) << std::endl;
        std::cout << "Success rate: " << (100.0 * passed_tests / total_tests) << "%" << std::endl;
        
        if (passed_tests == total_tests) {
            std::cout << "✅ All C++ KATs passed!" << std::endl;
        } else {
            std::cout << "❌ Some C++ KATs failed!" << std::endl;
        }
    }
};

int main() {
    GeometricWaveformKAT kat;
    kat.runAllTests();
    return 0;
}